### PR TITLE
[JENKINS-64479] ensure all logging is async to avoid deadlocks

### PR DESCRIPTION
**Pull request by jtnord**
[JENKINS-64479](https://issues.jenkins.io/browse/JENKINS-64479)

Logging during classloading is a surefire way to deadlock.
but we still want to have logs and in some cases we also want those logs
to be in the same place as other logs.

So introduce a new helper that punts the logging to a new Thread so that
we do not deadlock.

alternative to #17 based on suggestion by @jglick

### CODE REVIEW

File: bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            record.setParameters(parameters);
+            populateStackAndLogAsync(log, record);
+        }
+    }
+
+    private static void populateStackAndLogAsync(final Logger log, final LogRecord record) {
+        // call to make sure the stack is populated.
+        StackTraceElement[] stack = Thread.currentThread().getStackTrace();
+        if (stack.length > 3) {
+            // [0] is Thrad.getStackTrace
```
Comment (jglick) [inline bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java:57]:
```suggestion
            // [0] is Thread.getStackTrace
```

File: bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            // [1] is this method
+            // [2] is the caller (still this class)
+            // [3[ is the actual consumer we want
```
Comment (jglick) [inline bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java:60]:
```suggestion
            // [3] is the actual consumer we want
```

File: bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            StackTraceElement caller = stack[3];
+            record.setSourceClassName(caller.getClassName());
+            record.setSourceMethodName(caller.getMethodName());
+        }
+        new Thread(() -> log.log(record), "bytecode-compatability-transformer async logging").start();
```
Comment (jglick) [inline bytecode-compatibility-transformer/src/main/java/org/jenkinsci/bytecode/helper/LoggingHelper.java:65]:
```suggestion
        new Thread(() -> log.log(record), "bytecode-compatibility-transformer async logging").start();
```

**Review by jglick — APPROVED** [review]
As before I presume we have no way of testing that this works.

**Comment by jtnord** [comment]
> As before I presume we have no way of testing that this works.

testing strategy will involve building a snapshot of this,
building a snapshot of jenkins core using the just built version

enable FINEST logging on everything
perform several startups.  (with lots of plugins already installed)

it *may* be possible in a unit test by using creating our own ClassLoaders (or having some loading from Jars) and transforming the classes whilst also having some logging, but I am not convinced that is worth the investigation effort in this case.

**Review by amuniz — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
Not sure about new thread for every log entry. Project Loom is far ahead of our Java version, and FINEST logging is likely to be quite heavy. But it should address the issue indeed

**Review by res0nance — APPROVED** [review]

**Comment by jtnord** [comment]
> Not sure about new thread for every log entry. Project Loom is far ahead of our Java version, and FINEST logging is likely to be quite heavy. But it should address the issue indeed

I agree, however creating a Threadpool and a queue/drainer seems to be overkill here.
the alternative approach in #17 is an option, but not having the logs in the same place is also not ideal.

**Comment by amuniz** [comment]
@jtnord can we get this merged (I have no rights to do it)?

I've tested the change locally by:
* Build this (after fixing conflicts)
* Use it in jenkins-core
* Enable `FINEST` login on `org.*, com.*, hudson.*`
* Restart many times in a row
* Checked that there were no deadlocks
* The UI is responsive (slow because of all the logging, but ok)

Tried to push the conflicts resolution but I can push to your branch.

**Comment by jtnord** [comment]
@amuniz you can always doa PR to my branch.  but I have given you access to my fork so you can push the fixes.

**Comment by amuniz** [comment]
This was merged but the release didn't finish correctly because doesn't have permissions to upload to Artifactory. @timja you have cut a release recently, could you do another one please?

**Comment by timja** [comment]
released, fyi tests fail on java 11 😢

**Comment by amuniz** [comment]
> fyi tests fail on java 11

Ouch... I guess that's a problem then. This library is embedded in Jenkins core, so if there is an issue with Java 11 then Jenkins has an issue with Java 11.

Is this PR the one triggering the test failure?

**Comment by jtnord** [comment]
> test fail on Java 11

image

**Comment by jtnord** [comment]
I do see a compilation error on java11 though...  (after running clean) 

```
[INFO] Compiling 14 source files to C:\workarea\source\github\jenkinsci\bytecode-compatibility-transformer\bytecode-compatibility-transformer\target\classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] java.nio.file.NoSuchFileException: C:\workarea\source\github\jenkinsci\bytecode-compatibility-transformer\bytecode-compatibility-transformer\target\classes\META-INF\annotations\org.jenkinsci.bytecode.AdapterAnnotation
[INFO] 1 error
[INFO] -------------------------------------------------------------
```
Looks like an issue with the annotation creation not the runtime?

**Comment by jtnord** [comment]
FTR same occurs if I checkout `bct-parent-pom-2.0-beta-2` so this is not a new issue either with this PR or the unshaded ASM by the looks of it.

**Comment by jtnord** [comment]
@timja did you see specific test failures, or was it just the release failed?

I am (more) concerned if we see different things

**Comment by jtnord** [comment]
the following at least fixes the compilation failure.

```
diff --git a/bytecode-compatibility-transformer/pom.xml b/bytecode-compatibility-transformer/pom.xml
index 6afb791..c04f5cc 100644
--- a/bytecode-compatibility-transformer/pom.xml
+++ b/bytecode-compatibility-transformer/pom.xml
@@ -43,7 +43,7 @@
     <dependency>
       <groupId>org.jenkins-ci</groupId>
       <artifactId>annotation-indexer</artifactId>
-      <version>1.4</version>
+      <version>1.14</version>
     </dependency>
     <dependency>
       <groupId>org.ow2.asm</groupId>
```

I do not have time to look into this further this week, @amuniz are you able to take a look?

**Comment by timja** [comment]
Yeah the same message that @jtnord posted is what I got on the initial release attempt, swapped over to java 8 to release but not ideal.

**Comment by amuniz** [comment]
> @amuniz are you able to take a look?

I'll try, no promises until Monday though.

**Comment by jtnord** [comment]
erm....    https://github.com/jenkinsci/jenkins/pull/5526

I guess we can just go 🤷

**Comment by amuniz** [comment]
Oh! That's good actually!

**Pull request merged**